### PR TITLE
[wip] [Deprecate THCState] Moves p2pAccessEnabled to ATen

### PR DESCRIPTION
**Pull request by syed-ahmed**
This PR moves `p2pAccessEnabled` table from THCState to ATen, and hence works towards removing THCState. Previously, we were maintaining a lookup table of p2p connections. These connections were established during cuda init in groups of 8. That is, for instance, if you have a 16 gpu system and assuming all devices can access each other via p2p in that system, the current code sets up p2p connections between {0,1,2...7} and {8,9,10,..15} during lazy initialization. if a user want a connection from 1 to 10 let's say, and even though a p2p connection from 1 to 10 might be possible, the current code would not do a P2P memcpy, since Copy.cu reads from the p2pAccessEnabled table established at initialization. 

In the following figure, left is a copy from device 0 to 10, and right is from 0 to 1, which shows that we can't do p2p memcpy from 0 to 10, due to the connections being established in groups of 8. 
image 1

The current max number of simultaneous p2p connections that a device can have is 8 and even though if this number was increased in the future, our Copy.cu code would not allow someone from copying to devices outside their groups with P2P memcpy (since we use the conditional p2p_enabled in there which reads from the global p2pAccessEnabled table).

In this PR, rather than maintaining a lookup table, I created a CUDAP2PState struct, which establishes a p2p connection during construction and releases the connection when the connection is no longer needed. I have found that there is no significant overhead of not caching if a connection was established between two devices.

### CODE REVIEW

**Comment by ezyang** [comment]
> I have found that there is no significant overhead of not caching if a connection was established between two devices.

How did you do this performance experiment?

File: aten/src/ATen/cuda/CUDAContext.h [hunk 1]
```diff
@@ -36,6 +36,65 @@ It is expected that the modules whose functions compose this interface will
...
+  CUDAP2PState(int64_t src, int64_t target) 
+  : src_device(src), target_device(target) {
+    auto num_gpus = c10::cuda::device_count();
+    AT_ASSERT(src_device >= 0 && src_device < num_gpus);
+    AT_ASSERT(target_device >= 0 && target_device < num_gpus);
+    // if same device, no need for enabling
+    if(src_device == target_device) {
+      p2p_enable_status = true;
+    // if existing connection there, no need for enabling
+    } else if (!(p2p_enable_status)) {
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAContext.h]: Isn't this guaranteed to be true at construction

File: aten/src/ATen/cuda/CUDAContext.h [hunk 2]
```diff
@@ -36,6 +36,65 @@ It is expected that the modules whose functions compose this interface will
...
+      auto current_device = c10::cuda::current_device();
+      c10::cuda::set_device(src_device);
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAContext.h]: Use CUDAGuard!

File: aten/src/ATen/cuda/CUDAContext.h [hunk 3]
```diff
@@ -36,6 +36,65 @@ It is expected that the modules whose functions compose this interface will
...
+
+      int canAccess;
+      AT_CUDA_CHECK(cudaDeviceCanAccessPeer(&canAccess, src_device, target_device));
+      if (canAccess) {
+        cudaError_t err = cudaDeviceEnablePeerAccess(target_device, 0);
+        if (err == cudaErrorPeerAccessAlreadyEnabled || err == cudaSuccess) {
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAContext.h]: This looks very dodgy. Suppose two threads create CUDAP2PState(0,1). They will both attempt to enable the P2P access. Fine, you've written code to work around it. Then on destruction of the first state, peer access will be disabled, EVEN if another thread thinks it still wants it.

**Review by ezyang — CHANGES_REQUESTED** [review]
Unfortunately this design is not reentrant. Can't accept.

File: aten/src/ATen/cuda/CUDAContext.h [hunk 4]
```diff
@@ -36,6 +36,65 @@ It is expected that the modules whose functions compose this interface will
...
+          p2p_enable_status = true;
+        } else {
+          // if a device has more than 8 connections simultaneously (current max limit), we should
+          // get a cudaErrorTooManyPeers
```
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/cuda/CUDAContext.h]: No, it did not. It is lifted on the explorer systems. Docs erroneously don't list cudaErrorTooManyPeers as a possible return from cudaDeviceEnablePeerAccess, but the comment still says "Each device can support a system-wide maximum of eight peer connections." which is true for all systems except explorer. https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__PEER.html#group__CUDART__PEER_1g2b0adabf90db37e5cfddc92cbb2589f3

**Comment by syed-ahmed** [comment]
@ezyang Thanks for the review! I'll soon followup on it.

**Pull request closed without merging**